File: 198. House Robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0){
            return 0;
        }
        int length = nums.size();
        vector<int> dp(length + 1, 0);
        dp[0] = 0;
        dp[1] = nums[0];
        for(int i = 2; i <= length; i++){
            if(dp[i - 1] > dp[i - 2] + nums[i - 1]){
                dp[i] = dp[i - 1];
            }else{
                dp[i] = dp[i - 2] + nums[i - 1];
            }
        }
        return dp[length];
    }
};

File: 198. House Robber.java
class Solution {
    public int rob(int[] nums) {
        int[] dp = new int[2];
        dp[0] = nums[0];
        if(nums.length == 1) {
            return dp[0];
        }
        dp[1] = Math.max(nums[0], nums[1]);
        for(int i = 2; i < nums.length; i++) {
            int temp = dp[1];
            dp[1] = Math.max(dp[1], dp[0] + nums[i]);
            dp[0] = temp;
        }
        return dp[1];
    }
}